src/compiler/checker.ts

```diff
@@ -4579,17 +4684,33 @@ namespace ts {
                     const typeParameterNodes = index === (chain.length - 1) ? overrideTypeArguments : lookupTypeParameterNodes(chain, index, context);
                     const symbol = chain[index];
 
+                    const parent = chain[index - 1];
+                    let symbolName: string | undefined;
                     if (index === 0) {
                         context.flags |= NodeBuilderFlags.InInitialEntityName;
-                    }
-                    const symbolName = getNameOfSymbolAsWritten(symbol, context);
-                    context.approximateLength += symbolName.length + 1;
-                    if (index === 0) {
+                        symbolName = getNameOfSymbolAsWritten(symbol, context);
+                        context.approximateLength += (symbolName ? symbolName.length : 0) + 1;
                         context.flags ^= NodeBuilderFlags.InInitialEntityName;
                     }
+                    else {
+                        if (parent && getExportsOfSymbol(parent)) {
+                            const exports = getExportsOfSymbol(parent);
+                            forEachEntry(exports, (ex, name) => {
+                                if (getSymbolIfSameReference(ex, symbol) && !isLateBoundName(name) && name !== InternalSymbolName.ExportEquals) {
+                                    symbolName = unescapeLeadingUnderscores(name);
+                                    return true;
+                                }
+                            });
+                        }
+                    }
+                    if (!symbolName) {
+                        symbolName = getNameOfSymbolAsWritten(symbol, context);
+                    }
+                    context.approximateLength += symbolName.length + 1;
 
-                    const parent = chain[index - 1];
-                    if (!(context.flags & NodeBuilderFlags.ForbidIndexedAccessSymbolReferences) && parent && getMembersOfSymbol(parent) && getMembersOfSymbol(parent).get(symbol.escapedName) === symbol) {
+                    if (!(context.flags & NodeBuilderFlags.ForbidIndexedAccessSymbolReferences) && parent &&
+                        getMembersOfSymbol(parent) && getMembersOfSymbol(parent).get(symbol.escapedName) &&

```

Comment:
Hmm, do you think the VM is smart enough to keep up here? `getMembersOfSymbol` is called a bunch, and has a few conditionals+caching inside a second call. Maybe worth extracting a temp?

Reply:
`getMembersOfSymbol` looks up a value in a cache, so..... I hope so. It's an ugly place to extract a temp for perf, were it needed.